Sources/Common/Core/DataArray/index.js

```diff
@@ -313,42 +319,57 @@ function vtkDataArray(publicAPI, model) {
 // Object factory
 // ----------------------------------------------------------------------------
 
-const DEFAULT_VALUES = {
-  name: '',
-  numberOfComponents: 1,
-  size: 0,
-  dataType: DefaultDataType,
-  rangeTuple: [0, 0],
-  // values: null,
-  // ranges: null,
-};
+function defaultValues(initialValues) {
+  return {
+    name: '',
+    numberOfComponents: 1,
+    size: 0,
+    dataType: DefaultDataType,
+    rangeTuple: [0, 0],
+    // values: macro.newTypedArray(DefaultValues),
+    // ranges: null,
+    ...initialValues,
+  };
+}
 
 // ----------------------------------------------------------------------------
 
 export function extend(publicAPI, model, initialValues = {}) {
-  Object.assign(model, DEFAULT_VALUES, initialValues);
-
-  if (!model.empty && !model.values && !model.size) {
+  if (!initialValues.empty && !initialValues.values && !initialValues.size) {
     throw new TypeError(
       'Cannot create vtkDataArray object without: size > 0, values'
     );
   }
-
-  if (!model.values) {
-    model.values = macro.newTypedArray(model.dataType, model.size);
-  } else if (Array.isArray(model.values)) {
-    model.values = macro.newTypedArrayFrom(model.dataType, model.values);
-  }
-
-  if (model.values) {
-    model.size = model.values.length;
-    model.dataType = getDataType(model.values);
-  }
+  delete initialValues.empty;
+  Object.assign(initialValues, defaultValues(initialValues));
 
   // Object methods
   macro.obj(publicAPI, model);
   macro.set(publicAPI, model, ['name', 'numberOfComponents']);
 
+  model.dataType = initialValues.dataType
+    ? initialValues.dataType
+    : DefaultDataType;
+  delete initialValues.dataType;
+  if (!initialValues.values) {
+    if (!initialValues.size) initialValues.values = null;
+    else
+      initialValues.values = macro.newTypedArray(
+        model.dataType,
+        initialValues.size
+      );
+  } else if (Array.isArray(initialValues.values)) {
+    initialValues.values = macro.newTypedArrayFrom(
+      model.dataType,
+      initialValues.values
+    );
+  }
+
+  // We store the values given in a field called data in order to call setData
+  // lastly in the constructor
```

Comment:
Wait what? `setData(values, nbComponents)` is not the same thing as `values`

Reply:
As `setData` is the proper way to set `values` and `numberOfComponents`, it should be called when the DataArray is created. When calling `publicAPI.set(initialValues)` in the constructor, the `set` function looks for setters with the format `setMyPropName` and call them if they exist. Thus in order to call `setData` when creating the DataArray a property `data` should exist in the object `initialValues`. That's why @finetjul proposed to change `values` to `data` when calling `newInstance`. However, when copying an object through the call of `vtk(obj)`, the function calls the `constructor` method with the given properties of the object hence the model, which in the case of the DataArray contains a property named `values`. The `extend` function of the DataArray must therefore expect a property named values being given while still calling `setData`.

Reply:
Normally by the time `publicAPI.set(initialValues)` get called any "processed" data should be removed from initialValues which means you don't have any problem if you clean/remove any field that is getting processed in the `extend` function...

Reply:
The idea is to let `publicAPI.set(initialValues)` process the "data/values" initial parameter (by calling `setData()` on it) instead of handling it in the `extend()` function.

Reply:
In general when we have a 1 to 1 mapping with the data that is passed in, I would agree. But for something like that, I'm not sure. Are you aiming to do `vtkDataArray.newInstance({ data: [values, 3]})` ?

Reply:
After discussion, `setData()` will be called in `extend()`